File: eBPF_Supermarket/CPU_Subsystem/old_project/BCC_sar/src/wakeup/wakeup.py
#!/usr/bin/python3

from __future__ import print_function
import datetime
from os import getpid
from time import sleep, time
from bcc import BPF, PerfType, PerfSWConfig
import bpfutil, ctypes, argparse
import json
import sys
import os
import psutil

from wakeup_utils import get_sleep_func, deltaTimeMgr, MetricsAverager

BEGIN_SLEEP = 1
END_SLEEP = 2
BEGIN_WAIT = 3
END_WAIT = 4
BEGIN_RUN = 5
END_RUN = 6

def parse_args():
    parser = argparse.ArgumentParser(description="Wakeup Monitoring System", formatter_class=argparse.RawTextHelpFormatter)
    parser.add_argument("-t", "--type", default="time", help=
    """Type of output method.
1. event: Write event to file. (May cause lost of information)
2. time: print time of sleep,run,wait and compares their addition with actual time.
        -- Since our eBPF program is event driven, the static process like pid 1(systemd) 
        which don't cause any event won't be recorded, so their value may become less.
3. flame: Output Flame Format TextFile to Produce frame SVG.
    """)
    parser.add_argument("-p", "--pid", default=0, type=int, 
        help="The Process needs to be attached on.")
    parser.add_argument("--time", default=3, type=float, help="excution time")
    parser.add_argument("--filter", default="Begin_Run", type=str, help="Type of event to filter.")
    parser.add_argument("-o", "--output", default="event.txt", type=str, 
        help="The output file which saves all the content.")

    return parser.parse_args()


if __name__ == "__main__":
    args = parse_args()

    bpf = BPF(src_file="wakeup.c")
    bpf.attach_tracepoint(tp="sched:sched_wakeup", fn_name="trace_wakeup")
    bpf.attach_tracepoint(tp="sched:sched_switch", fn_name="trace_sched_switch")

    # 分别挂载在一个函数的进入和退出
    # bpf.attach_kprobe(event="try_to_wake_up", fn_name="kprobe_try_to_wake_up")
    # bpf.attach_kretprobe(event="try_to_wake_up", fn_name="kretprobe_try_to_wake_up")

    # 传递给bpf程序的定时事件
    bpf.attach_perf_event(ev_type=PerfType.SOFTWARE,
        ev_config=PerfSWConfig.CPU_CLOCK,
        fn_name="tick_update",
        sample_period=0, 
        sample_freq=3
    )

    # 内核中的重要变量定义，留作参考，勿删！
    # pcpu_base_addr = int(bpfutil.find_ksym("pcpu_base_addr"), base=16) # u32, Read Only
    # __per_cpu_start = int(bpfutil.find_ksym("__per_cpu_start"), base=16) # u32, Read Only
    # pcpu_unit_offsets = int(bpfutil.find_ksym("pcpu_unit_offsets"), base=16) # u32 array, Read Only

    # cpu偏移量
    __per_cpu_offset = int(bpfutil.find_ksym("__per_cpu_offset"), base=16) # u32 array, Inited
    # runqueues, 运行队列的percpu变量
    runqueues = int(bpfutil.find_ksym("runqueues"), base=16) # per_cpu var, u32
    __preempt_count = int(bpfutil.find_ksym("__preempt_count"), base=16) # int, 32bit

    bpf['symAddr'][0] = ctypes.c_ulonglong(__per_cpu_offset)
    bpf['symAddr'][1] = ctypes.c_ulonglong(runqueues)
    bpf['symAddr'][2] = ctypes.c_ulonglong(__preempt_count) # 抢占标志位
    # 计算方式：某个cpu的PERCPU变量的地址 = PERCPU变量的地址 + __per_cpu_offset[cpu]

    # 传递自己的PID，使BPF避开这个pid，这件事非常重要！
    bpf["ownPid"][0] = ctypes.c_int(getpid())

    target_pid = args.pid
    nowtime = 0

    '''
    # 把栈信息写入到文件中
    def onexit():
        with open("sleep_stack.txt", "w") as f:
            for pid in pidSleepMap.keys():
                f.write("pid {}:\n".format(pid))
                for str in pidSleepMap[pid]:
                    f.write(str)
                    f.write("\n")
                f.write("\n")

        # 诀窍：读取schedule栈后(可用含有schedule来表示)的第一个函数
        with open("stack_type.txt", "w") as f:
            for stack in stack_set:
                f.write(stack)
                f.write("\n")

        print("Already Write.")
    atexit.register(onexit)
    '''

    # type=time: 打印某个进程在运行时的运行、等待、睡眠事件占比
    if args.type == "time":
        thead = r"%runTime  %waitTime  %sleepTime  allTime/ms  elapsed/ms  sleepTime/ms  runTime/us  runslice/us  runCount"
        print(thead)
        deltaMgr = deltaTimeMgr(4)

        sleepTime = MetricsAverager()
        runTime = MetricsAverager()
        runsliceTime = MetricsAverager()
        runCount = MetricsAverager()

        sleepStart = {}
        runStart = {}
        runTimeMap = {}
        def metric_record(cpu, data, size):
            global sleepStart, runStart, runTimeMap
            global sleepTime, runTime, runsliceTime, runCount
            # 成员：pid, comm, type, stackid, waker, waker_comm, time
            event = bpf["events"].event(data)

            pid = event.pid
            if pid != args.pid: return

            # comm = event.comm.decode('utf-8')
            # waker_comm = event.waker_comm.decode('utf-8')

            if event.type == BEGIN_SLEEP:
                # wchan = get_sleep_func(bpf, event.stackid)
                sleepStart[pid] = event.time

                if pid in runTimeMap:
                    count = len(runTimeMap[pid])
                    sum_time = 0
                    for i in runTimeMap[pid]:
                        sum_time += i
                    runCount.add(count)
                    runTime.add(sum_time)
                    runTimeMap.clear()

            elif event.type == END_SLEEP:
                if pid not in sleepStart: return
                time = (event.time - sleepStart[pid]) / 1000 # us
                del sleepStart[pid] # 删除键pid，表示pid已不在睡眠状态中了
                sleepTime.add(time)
            
            elif event.type == BEGIN_RUN:
                runStart[pid] = event.time

            elif event.type == END_RUN:
                if pid not in runStart: return
                time = (event.time - runStart[pid]) / 1000
                runsliceTime.add(time)

                if pid not in runTimeMap:
                    runTimeMap[pid] = []

                runTimeMap[pid].append(time)

        bpf["events"].open_perf_buffer(metric_record, page_cnt=128)
        while 1:
            runlast = waitlast = sleeplast = 0

            if ctypes.c_uint32(target_pid) in bpf["runlast"]: 
                runlast = bpf['runlast'][ctypes.c_uint32(target_pid)].value
            if ctypes.c_uint32(target_pid) in bpf["waitlast"]: 
                waitlast = bpf['waitlast'][ctypes.c_uint32(target_pid)].value
            if ctypes.c_uint32(target_pid) in bpf["sleeplast"]: 
                sleeplast = bpf['sleeplast'][ctypes.c_uint32(target_pid)].value
            all = runlast + waitlast + sleeplast

            runlast, waitlast, sleeplast, all = deltaMgr.update([runlast, waitlast, sleeplast, all])

            if all != 0:
                # sleepTime/ms  runTime/us  runslice/us  runCount
                print("%8.2f  %9.2f  %9.2f  %10d  %10d  %12.1f  %10.1f  %12.1f  %8.1f" % 
                    (runlast / all * 100, waitlast / all * 100, sleeplast / all * 100,
                    int(all / 1000000), 1000, 
                    sleepTime.recent()/1000, runTime.recent(), 
                    runsliceTime.recent(), runCount.recent() ) )
            else: # 避免除以0
                print("%8.2f  %9.2f  %9.2f  %10d  %10d  %12.1f  %10.1f  %12.1f  %8.1f" % 
                    (0, 0, 0, int(all / 1000000), 1000,
                    sleepTime.recent()/1000, runTime.recent(), 
                    runsliceTime.recent(), runCount.recent() ) )
            
            bpf.perf_buffer_poll()
            sleep(1)
            nowtime += 1
            if nowtime % 12 == 0:
                print()
                print(thead)

    # type=event: 生成事件日志，包括运行、等待、睡眠
    elif args.type == 'event':
        pidEvent = {} # 存储每个pid的event列表
        f = open(args.output, "w")

        # 记录type序号对应的事件类型
        typeNameMap = {
            1: "Begin_Sleep",
            2: "End_Sleep",
            3: "Begin_Wait",
            4: "End_Wait",
            5: "Begin_Run",
            6: "End_Run"
        }

        # perf数据接收回调函数
        def print_event(cpu, data, size):
            global cnt
            cnt += 1
            # 成员：pid, comm, type, stackid, waker, waker_comm, time
            event = bpf["events"].event(data)

            pid = event.pid
            stackid = event.stackid

            comm = event.comm.decode('utf-8')
            waker_comm = event.waker_comm.decode('utf-8')

            # 若当前类型为BeginSleep或者EndSleep，读取栈信息到stack_str中
            if event.type <= 2:
                # 对于BEGIN_SLEEP, 打印wchan
                stack_str = ""
                if event.type == BEGIN_SLEEP:
                    stack_str = get_sleep_func(bpf, stackid)
                else: # END_SLEEP
                    # 之后可能做一下数据分类，现在先只用第一项
                    for addr in bpf["stacktraces"].walk(stackid):
                        sym = bpf.ksym(addr).decode('utf-8', 'replace')
                        # 采集整个栈的信息
                        # stack_str = stack_str + "\t" + sym + "\n"
                        stack_str = sym
                        break
            else:
                stack_str = "None"

            if pid not in pidEvent:
                pidEvent[pid] = []

            # if typeNameMap[event.type] not in ["End_Sleep", "End_Run", "End_Wait"]:
            try:
                pid_str = "#%d pid:%d %s type:%s waker:%d %s time:%dus\n%s\n\n" \
                    % (cnt, pid, comm, typeNameMap[event.type], 
                    event.waker, waker_comm, int(event.time / 1000), 
                    stack_str)
            except Exception as e:
                print(e)
                os._exit(0)

            # 如果配置了pid项，那么只打印与该pid有关的事件
            if args.pid != 0:
                if pid == args.pid:
                    f.write(pid_str)
            else:
                f.write(pid_str)
            
            pidEvent[pid].append(pid_str)

        # 准备从perf_output中接收数据
        start = time()
        bpf["events"].open_perf_buffer(print_event, page_cnt=128)
        global cnt
        cnt = 0

        while 1:
            bpf.perf_buffer_poll()
            sleep(1)
            if time() - start > args.time:
                break

            # nowtime += 1
            # print(nowtime)
            # if nowtime == 10:
            #     break

        f.close() # 关闭文件
        print("EVENT cnt = ", cnt)
        print("Tracepoint count = ", bpf["countMap"][0].value)

    # type=flame: 生成火焰图所需要的格式化文件
    elif args.type == "flame":
        f = open(args.output, "w")
        runbegin = {}
        runEnd = {}

        # 检查是否设置专门监测某一个pid
        pid_configured = (args.pid != 0)

        if pid_configured:
            print("只跟踪pid = {}进程的活动".format(args.pid))

        def flame_record(cpu, data, size):
            # 成员：pid, comm, type, stackid, waker, waker_comm, time
            event = bpf["events"].event(data)

            pid = event.pid
            comm = event.comm.decode('utf-8')
            # waker_comm = event.waker_comm.decode('utf-8')

            if cpu != 0: return

            if event.type == BEGIN_RUN:
                runbegin[pid] = event.time
                if pid_configured and pid == args.pid and pid in runEnd:
                    time = int((event.time - runEnd[pid]) / 1000)
                    f.write("{0};{1} {2}\n".format(comm, "stop", time))

            # 如果配置了pid项，那么只输出对应pid的所有事件
            elif event.type == END_RUN and pid in runbegin:
                if pid != 0:
                    time = int((event.time - runbegin[pid]) / 1000) # us
                else:
                    time = 3 # idle进程的运行时间一律设为3个单位，以优化显示

                if (pid_configured and pid == args.pid) or (not pid_configured):
                    f.write("{0};{1} {2}\n".format(comm, pid, time))

                runEnd[pid] = event.time

        bpf["events"].open_perf_buffer(flame_record, page_cnt=128)
        start = time()
        while 1:
            bpf.perf_buffer_poll()
            sleep(1)
            if time() - start > args.time:
                break

    # type=lifeline：生成某个进程的生命周期数据，只包括运行和睡眠
    elif args.type == "lifeline":
        f = open(args.output, "w")
        sleep_start = -1
        run_start = -1
        print("跟踪pid = {}进程的生命周期...".format(args.pid))

        def flame_record(cpu, data, size):
            global sleep_start, run_start
            # 成员：pid, comm, type, stackid, waker, waker_comm, time
            event = bpf["events"].event(data)

            pid = event.pid
            if pid != args.pid: return

            comm = event.comm.decode('utf-8')
            waker_comm = event.waker_comm.decode('utf-8')

            if event.type == BEGIN_SLEEP:
                wchan = get_sleep_func(bpf, event.stackid)
                # BPF 获取到的timestamp实际上是系统的启动时间，先留置
                # 之后再转换为时分秒
                sleep_start = event.time
                f.write("{}: [cpu{}] {}-{} Begin_Sleep on {}\n".format(event.time, 
                    cpu, pid, comm, wchan))

            elif event.type == END_SLEEP:
                if sleep_start != -1:
                    time = ((event.time - sleep_start) / 1000000)
                else:
                    time = -1

                stack_str = ""
                for addr in bpf["stacktraces"].walk(event.stackid):
                    sym = bpf.ksym(addr).decode('utf-8', 'replace')
                    # 采集整个栈的信息
                    stack_str = stack_str + "\t" + sym + "\n"
                    # break

                f.write("%d: [cpu%d] %d-%s End_Sleep lasting %.1fms (wake by %d-%s, context %d)\n%s\n"
                    % (event.time, cpu, pid, comm, time, 
                    event.waker, waker_comm, event.preempt_count, stack_str))
            
            elif event.type == BEGIN_RUN:
                run_start = event.time

            elif event.type == END_RUN:
                if run_start != -1:
                    time = int((event.time - run_start) / 1000) # us
                else:
                    time = -1
                f.write("%d: [cpu%d] %d-%s Run for %dus\n" % (event.time, cpu,
                    pid, comm, time))

        bpf["events"].open_perf_buffer(flame_record, page_cnt=128)
        start = time()
        while 1:
            bpf.perf_buffer_poll()
            sleep(1)
            if time() - start > args.time:
                break

    # 生成进程的状态切换JSON文件，写入到python_exporter目录下
    elif args.type == "json":
        print("Monitoring process pid = %d" % args.pid)
        f = open("../../exporter/metrics.json", "w") # 写死的输出路径，之后可以改
        waker_comm = ""
        jsonMap = []
        boot_time = psutil.boot_time()  # 系统启动时间戳
        curState = "run"

        def flame_record(cpu, data, size):
            global waker_comm, curState
            # 成员：pid, comm, type, stackid, waker, waker_comm, time
            event = bpf["events"].event(data)
            if event.pid != args.pid: return
            # event.time是相对于系统启动时间的时间戳

            nowtime = boot_time + event.time / 1000000000

            if event.type == BEGIN_RUN:
                curState = "run"
                jsonMap.append({"time": nowtime, "state": "run-cpu%d" % cpu, "pid": args.pid, "evt_time": event.time})

            elif event.type == BEGIN_SLEEP:
                curState = "sleep"
                wchan = get_sleep_func(bpf, event.stackid)
                jsonMap.append({"time": nowtime, "state": "sleep-%s" % (wchan,), "pid": args.pid, "evt_time": event.time})

            elif event.type == BEGIN_WAIT:
                info = "" if curState == "run" else waker_comm
                jsonMap.append({"time": nowtime, "state": "wait-%s" % (info,), "pid": args.pid, "evt_time": event.time})

            elif event.type == END_SLEEP:
                waker_comm = event.waker_comm.decode('utf-8')

        bpf["events"].open_perf_buffer(flame_record, page_cnt=128)
        start = time()
        while 1:
            bpf.perf_buffer_poll()
            sleep(1)
            if time() - start > args.time:
                break

        json_str = json.dumps(jsonMap)
        f.write(json_str)
        f.close()
        print("Produce %d Records." % len(jsonMap))
    
    '''
    # perf缓冲区的用法(previous)
    def print_event(cpu, data, size):
        event = bpf["events"].event(data)
        print("recv msg.")

    # loop with callback to print_event
    bpf["events"].open_perf_buffer(print_event)

    while 1:
        bpf.perf_buffer_poll() # 仅适用于与poll不发生牵连影响的perf_submit
    '''

File: eBPF_Supermarket/CPU_Subsystem/old_project/BCC_sar/src/wakeup/wakeup.c
#include <uapi/linux/ptrace.h>
#include <linux/sched.h>

#define PRINT_SLEEP 1
#define TARGET_PID 10327

BPF_ARRAY(symAddr, u64, 5);
BPF_ARRAY(countMap, u64, 1);
BPF_ARRAY(ownPid, int, 1);
BPF_STACK_TRACE(stacktraces, 50009);

struct sleep_record {
    u32 state;
    u32 in_sleep;
    u64 start;
};

BPF_HASH(sleepBegin, int, struct sleep_record, 4096);
BPF_HASH(sleeplast, int, u64, 4096);
BPF_HASH(runlast, int, u64, 4096);
BPF_HASH(runBegin, int, u64, 4096);
BPF_HASH(waitlast, int, u64, 4096);
BPF_HASH(waitBegin, int, u64, 4096);

BPF_PERF_OUTPUT(events);

struct wakeup_struct {
    unsigned long long pad;
    char comm[16];
    int pid;
    int prio;
    int success;
    int target_cpu;
};

struct wait_struct {
    unsigned long long pad;
    char comm[16];
    int pid;
    int prio;
};

struct sleep_struct {
    unsigned long long pad;
    char comm[16];
    int pid;
    u64 delay;
};

struct rq {
	raw_spinlock_t lock;
	unsigned int nr_running;
	unsigned int nr_numa_running;
	unsigned int nr_preferred_running;
	unsigned int numa_migrate_on;
	long unsigned int last_blocked_load_update_tick;
	unsigned int has_blocked_load;
};


struct cswch_args {
	unsigned long long pad;
	char prev_comm[16];
	pid_t prev_pid;
	int prev_prio;
	long prev_state;
	char next_comm[16];
	pid_t next_pid;
	int next_prio;
};

#define TASK_RUNNING			0x0000
#define TASK_INTERRUPTIBLE		0x0001
#define TASK_UNINTERRUPTIBLE		0x0002


#define BEGIN_SLEEP 1
#define END_SLEEP 2
#define BEGIN_WAIT 3
#define END_WAIT 4
#define BEGIN_RUN 5
#define END_RUN 6

struct perf_data {
    int stackid; // 调用栈的信息: waker/sleeper
    int pid;
    char comm[16];
    int type;
    int waker;
    char waker_comm[16];
    u64 time;
    int preempt_count;
};

__always_inline static void inc_count() {
    u32 key = 0;
    u64 initval = 1, *valp = countMap.lookup(&key);
    if (valp) *valp += 1;
    else countMap.update(&key, &initval);
}

__always_inline static int getMyPid() {
    u32 key = 0;
    int *valp = ownPid.lookup(&key);
    if (valp) return *valp;
    else return -1;
}

__always_inline static void copy_comm(char *src, char *dst) {
    for (int i = 0; i < 15; i++) {
        dst[i] = src[i];
    }
    dst[15] = 0;
}

// 获取当前运行的cpu上抢占标志位的值
__always_inline static int get_preempt_count() {
    u64 *valp;
    u64 p_offset = 0, p_preempt = 0;

    u32 key = 0;
    valp = symAddr.lookup(&key);
    if(valp) p_offset = *valp;

    key = 2;
    valp = symAddr.lookup(&key);
    if(valp) p_preempt = *valp;

    u64 percpu_offset[2];
    bpf_probe_read_kernel(percpu_offset, sizeof(u64) * 2, p_offset);

    int preempt_count = 125868; // magic number 125868表示出错
    // 变量值在BPF中做下标时，需要限制范围
    u32 cpu = bpf_get_smp_processor_id();
    // bpf_trace_printk("cpu = %d\n", cpu);

    // May Be BPF's Problem? So I have to get percpu_offset in this stupid way.
    // tmp记录了percpu_offset(u64)的某个值，坚决不能用int或者u32，因为向下转型会出现问题！！
    u64 tmp;
    if (cpu == 1) {
        tmp = percpu_offset[1];
    }
    else {
        tmp = percpu_offset[0];
    }

    // p_preempt = p_preempt + tmp;
    bpf_probe_read_kernel(&preempt_count, sizeof(preempt_count), p_preempt + tmp);
    preempt_count = preempt_count & (0x7fffffff);
    return preempt_count;
}

// 事件统一用perf output形式传递到用户空间

// from trace_sched_switch
__always_inline static void begin_sleep(struct pt_regs *ctx) {
    struct task_struct *ts = bpf_get_current_task();
    u32 pid = ts->pid;
    struct perf_data data = {};

    data.stackid = stacktraces.get_stackid(ctx, 0);
    data.pid = pid;
    data.type = BEGIN_SLEEP;
    data.waker = 0;
    // u64 bpf_ktime_get_ns(void) Description Return the time elapsed since system boot, in nanoseconds.
    data.time = bpf_ktime_get_ns();
    data.preempt_count = 0;

    // 获取进程的comm
    bpf_get_current_comm(&data.comm, sizeof(data.comm));

    // 向用户空间传送睡眠进程名、pid以及发生阻塞时的栈信息
    events.perf_submit(ctx, &data, sizeof(data));
    // inc_count();
    // bpf_trace_printk("#%-8d Begin Sleep", pid);
}

// #<pid>表示主语为pid
__always_inline static void end_sleep(struct pt_regs *waker_ctx, 
                                      struct wakeup_struct *ws, int waker_pid) {
    struct perf_data data = {};
    struct task_struct *ts = bpf_get_current_task();

    data.stackid = stacktraces.get_stackid(waker_ctx, 0);
    data.pid = ws->pid;
    data.type = END_SLEEP;
    data.waker = waker_pid;
    bpf_probe_read_kernel(data.waker_comm, 16, ts->comm);
    data.comm[15] = 0;
    copy_comm(ws->comm, data.comm);
    data.time = bpf_ktime_get_ns();
    data.preempt_count = get_preempt_count();

    events.perf_submit(waker_ctx, &data, sizeof(data));
    // bpf_trace_printk("#%-8d Wakeup %d", waker_pid, ws->pid);
    // bpf_trace_printk("#%-8d End Sleep", ws->pid);
    // inc_count();
}

__always_inline static void begin_run(int pid, char *comm, struct pt_regs *ctx) {
    struct perf_data data = {};

    data.stackid = 0;
    data.pid = pid;
    data.type = BEGIN_RUN;
    data.waker = 0;
    copy_comm(comm, data.comm);
    data.time = bpf_ktime_get_ns();
    data.preempt_count = 0;

    events.perf_submit(ctx, &data, sizeof(data));
    // bpf_trace_printk("#%-8d Begin Run", pid);
    // inc_count();
}

__always_inline static void end_run(int pid, char *comm, struct pt_regs *ctx) {
    struct perf_data data = {};

    data.stackid = 0;
    data.pid = pid;
    data.type = END_RUN;
    data.waker = 0;
    copy_comm(comm, data.comm);
    data.time = bpf_ktime_get_ns();
    data.preempt_count = 0;

    events.perf_submit(ctx, &data, sizeof(data));
    // bpf_trace_printk("#%-8d End Run", pid);
    // inc_count();
}

__always_inline static void begin_wait(int pid, char *comm, struct pt_regs *ctx) {
    struct perf_data data = {};

    data.stackid = 0;
    data.pid = pid;
    data.type = BEGIN_WAIT;
    data.waker = 0;
    copy_comm(comm, data.comm);
    data.time = bpf_ktime_get_ns();
    data.preempt_count = 0;

    events.perf_submit(ctx, &data, sizeof(data));
    // bpf_trace_printk("#%-8d Begin Wait", pid);
    // inc_count();
}

__always_inline static void end_wait(int pid, char *comm, struct pt_regs *ctx) {
    struct perf_data data = {};

    data.stackid = 0;
    data.pid = pid;
    data.type = END_WAIT;
    data.waker = 0;
    copy_comm(comm, data.comm);
    data.time = bpf_ktime_get_ns();
    data.preempt_count = 0;

    events.perf_submit(ctx, &data, sizeof(data));
    // bpf_trace_printk("#%-8d End Wait", pid);
    // inc_count();
}

int trace_sched_switch(struct cswch_args *ctx) {
    int myPid = getMyPid();
    inc_count();
    if (ctx->next_pid == ctx->prev_pid) return 0;
    // if (ctx->next_pid == myPid || ctx->prev_pid == myPid) return 0;

    struct task_struct *ts = bpf_get_current_task();
    struct sleep_record sr;
    u32 next = ctx->next_pid, prev = ctx->prev_pid;
    u64 time = bpf_ktime_get_ns(), *valp, delta, initval = 0;

    begin_run(ctx->next_pid, ctx->next_comm, ctx);
    end_run(ctx->prev_pid, ctx->prev_comm, ctx);

    // state 在新版的linux内核中被改成了_state
    if ( (ts->__state & TASK_INTERRUPTIBLE) || (ts->__state & TASK_UNINTERRUPTIBLE) ) {
        // 进入睡眠状态
        sr.state = ts->__state;
        sr.in_sleep = 1;
        sr.start = time;
        sleepBegin.update(&prev, &sr);

        begin_sleep((struct pt_regs *)ctx);
    } else if (ts->__state == TASK_RUNNING) {
        // 开始等待
        waitBegin.update(&prev, &time);
        begin_wait(ctx->prev_pid, ctx->prev_comm, ctx);
    }

    // 更新新进程的开始执行时间
    runBegin.update(&next, &time);

    // // next: 结束睡眠(一般不可能。一般要先被wakeup，不可能直接被调度运行)
    // struct sleep_record *p_sr;
    // p_sr = sleepBegin.lookup(&next);
    // if (p_sr && p_sr->in_sleep) {
    //     u64 last = time - p_sr->start;
    //     bpf_trace_printk("end sleep %d us\n", last);
    // }

    // next: 结束等待
    end_wait(ctx->next_pid, ctx->next_comm, ctx);
    valp = waitBegin.lookup(&next);
    if (valp && *valp != 0) {
        // bpf_trace_printk("valp = %llu\n", *valp);
        delta = time - *valp;

        valp = waitlast.lookup(&next);
        if (valp) *valp += delta;
        else waitlast.update(&next, &delta);

        waitBegin.update(&next, &initval);
    }

    // 统计旧进程的执行时间
    valp = runBegin.lookup(&prev);
    if (valp) {
        delta = time - *valp;
        valp = runlast.lookup(&prev);
        if (valp) *valp += delta;
        else runlast.update(&prev, &delta);
    }
    
    return 0;
}


int trace_wakeup(struct wakeup_struct *ws) {
    inc_count();
    struct task_struct *ts = (struct task_struct *)bpf_get_current_task();
    int myPid = getMyPid();
    // if (ws->pid == myPid && ts->pid == myPid) return 0;

    // char comm[16];
    // bpf_probe_read_kernel(comm, 16, ts->comm);
    // comm[15] = 0;

    bpf_trace_printk("wakeup_preempt = %d\n", get_preempt_count());

    u32 pid = ws->pid;
    u64 *valp, nowtime = bpf_ktime_get_ns();
    struct sleep_record *p_sr = sleepBegin.lookup(&pid);
    if (p_sr && p_sr->in_sleep) {
        u64 time = nowtime - p_sr->start;
        // if (time > 1000000 && PRINT_SLEEP && pid == TARGET_PID) { // 1ms
        //     bpf_trace_printk("process %d sleep for %d us.(wakeup by %d)\n", pid, time / 1000, ts->pid);
        // }
        p_sr->in_sleep = 0;

        // 结束睡眠
        end_sleep((struct pt_regs *)ws, ws, ts->pid);

        // 开始等待
        waitBegin.update(&pid, &nowtime);
        begin_wait(pid, ws->comm, ws);

        valp = sleeplast.lookup(&pid);
        if (valp) *valp += time;
        else sleeplast.update(&pid, &time);
    }
    
    return 0;
}

// 在内核中，64位系统的 sizeof(long) == 8
int tick_update() {
    u64 *valp;
    u64 p_offset = 0, p_rq = 0;

    u32 key = 0;
    valp = symAddr.lookup(&key);
    if(valp) p_offset = *valp;

    key = 1;
    valp = symAddr.lookup(&key);
    if(valp) p_rq = *valp;

    u64 percpu_offset[2];
    struct rq _rq;
    bpf_probe_read_kernel(percpu_offset, sizeof(u64) * 2, p_offset);
    u64 rq_ptr = p_rq + percpu_offset[0]; // 默认读取CPU0的Rq
    bpf_probe_read_kernel(&_rq, sizeof(struct rq), rq_ptr);
    
    // int preempt_count = get_preempt_count();
    // bpf_trace_printk("preempt_count = %d\n", preempt_count);
    
    struct task_struct *ts = (struct task_struct *)bpf_get_current_task();
    int *pusage = &ts->usage; // usage表示进程被等待队列引用的数目
    // bpf_trace_printk("curr task usage = %d\n", *pusage);

    // 读取rq
    // bpf_trace_printk("%lx %lx\n", percpu_offset[0], percpu_offset[1]);
    // bpf_trace_printk("rqlen = %d\n", _rq.nr_running);
    return 0;
}


// 主要想评估一下是否wakeup成功，但是我觉得在trace_wake里面wake的程序应该都能成功
int kprobe_try_to_wake_up(struct pt_regs *ctx) {
    struct task_struct *p = PT_REGS_PARM1(ctx);
    struct task_struct *curr = bpf_get_current_task();
    bpf_trace_printk("try_to_wake_up %d, which state = %x", p->pid, p->__state);

    return 0;
}

// 获取返回值例程
int kretprobe_try_to_wake_up(struct pt_regs *ctx) {
    // 从ctx里读取返回值
    int ret = PT_REGS_RC(ctx);

    bpf_trace_printk("[%d]try_to_wake_up", ret);
    return 0;
}